File: app-linux/src/weather.rs
//! Weather collection per docs/sdd.md §7.4: geocoding + Open-Meteo forecast,
//! WMO weather_code → text mapping, parsed into `WeatherReport` (§5.4).
//! JSON parsing and the WMO mapping are pure functions, covered by
//! docs/test-vectors/weather-json/ and docs/test-vectors/wmo-codes/.

use std::fmt;
use std::time::Duration;

use serde_json::Value;

use crate::config::WeatherConfig;
use crate::model::{WeatherDay, WeatherNow, WeatherReport};

#[derive(Debug)]
pub enum WeatherError {
    Config(String),
    Network(String),
    Parse(String),
}

impl fmt::Display for WeatherError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            WeatherError::Config(msg) => write!(f, "weather config error: {msg}"),
            WeatherError::Network(msg) => write!(f, "weather network error: {msg}"),
            WeatherError::Parse(msg) => write!(f, "weather parse error: {msg}"),
        }
    }
}

impl std::error::Error for WeatherError {}

/// WMO weather_code → text mapping, per docs/sdd.md §10.2 (wmo-codes) /
/// docs/test-vectors/wmo-codes/mapping.json. Pure function.
pub fn wmo_description(code: i64) -> &'static str {
    match code {
        0 => "Clear sky",
        1 => "Mainly clear",
        2 => "Partly cloudy",
        3 => "Overcast",
        45 => "Fog",
        48 => "Depositing rime fog",
        51 => "Light drizzle",
        53 => "Moderate drizzle",
        55 => "Dense drizzle",
        56 => "Light freezing drizzle",
        57 => "Dense freezing drizzle",
        61 => "Slight rain",
        63 => "Moderate rain",
        65 => "Heavy rain",
        66 => "Light freezing rain",
        67 => "Heavy freezing rain",
        71 => "Slight snow fall",
        73 => "Moderate snow fall",
        75 => "Heavy snow fall",
        77 => "Snow grains",
        80 => "Slight rain showers",
        81 => "Moderate rain showers",
        82 => "Violent rain showers",
        85 => "Slight snow showers",
        86 => "Heavy snow showers",
        95 => "Thunderstorm",
        96 => "Thunderstorm with slight hail",
        99 => "Thunderstorm with heavy hail",
        _ => "Unknown",
    }
}

/// Parse an Open-Meteo `forecast` JSON response into a `WeatherReport`. Pure
/// function — covered by docs/test-vectors/weather-json/. `location_label`,
/// `temperature_unit` and `fetched_at` come from the caller (config/clock),
/// everything else is read from the JSON itself.
pub fn parse_forecast_json(
    json: &Value,
    location_label: &str,
    temperature_unit: &str,
    fetched_at: i64,
) -> Result<WeatherReport, WeatherError> {
    let latitude = json
        .get("latitude")
        .and_then(Value::as_f64)
        .ok_or_else(|| WeatherError::Parse("missing `latitude`".to_string()))?;
    let longitude = json
        .get("longitude")
        .and_then(Value::as_f64)
        .ok_or_else(|| WeatherError::Parse("missing `longitude`".to_string()))?;

    let current_obj = json
        .get("current")
        .ok_or_else(|| WeatherError::Parse("missing `current`".to_string()))?;
    let temperature = current_obj
        .get("temperature_2m")
        .and_then(Value::as_f64)
        .ok_or_else(|| WeatherError::Parse("missing `current.temperature_2m`".to_string()))?;
    let weather_code = current_obj
        .get("weather_code")
        .and_then(Value::as_i64)
        .ok_or_else(|| WeatherError::Parse("missing `current.weather_code`".to_string()))?;
    let current = WeatherNow {
        temperature,
        weather_code,
        description: wmo_description(weather_code).to_string(),
    };

    let mut daily = Vec::new();
    if let Some(daily_obj) = json.get("daily") {
        let dates = daily_obj
            .get("time")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default();
        let codes = daily_obj.get("weather_code").and_then(Value::as_array);
        let temp_maxes = daily_obj.get("temperature_2m_max").and_then(Value::as_array);
        let temp_mins = daily_obj.get("temperature_2m_min").and_then(Value::as_array);
        let precip = daily_obj
            .get("precipitation_probability_max")
            .and_then(Value::as_array);

        for (i, date_val) in dates.iter().enumerate() {
            let date = match date_val.as_str() {
                Some(d) => d.to_string(),
                None => continue,
            };
            let code = codes.and_then(|a| a.get(i)).and_then(Value::as_i64);
            let temp_max = temp_maxes.and_then(|a| a.get(i)).and_then(Value::as_f64);
            let temp_min = temp_mins.and_then(|a| a.get(i)).and_then(Value::as_f64);
            let (code, temp_max, temp_min) = match (code, temp_max, temp_min) {
                (Some(c), Some(mx), Some(mn)) => (c, mx, mn),
                // A day missing any required field is dropped rather than
                // failing the whole report, per SDD §7.4 degradation intent.
                _ => continue,
            };
            let precipitation_probability_max =
                precip.and_then(|a| a.get(i)).and_then(Value::as_i64);
            daily.push(WeatherDay {
                date,
                temp_max,
                temp_min,
                weather_code: code,
                description: wmo_description(code).to_string(),
                precipitation_probability_max,
            });
        }
    }

    Ok(WeatherReport {
        location_label: location_label.to_string(),
        latitude,
        longitude,
        fetched_at,
        current,
        daily,
        temperature_unit: temperature_unit.to_string(),
    })
}

fn now_unix() -> i64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

fn http_get_json(url: &str, timeout: Duration) -> Result<Value, WeatherError> {
    let config = ureq::Agent::config_builder()
        .timeout_global(Some(timeout))
        .build();
    let agent: ureq::Agent = config.into();
    let body = agent
        .get(url)
        .call()
        .map_err(|e| WeatherError::Network(e.to_string()))?
        .body_mut()
        .read_to_string()
        .map_err(|e| WeatherError::Network(e.to_string()))?;
    serde_json::from_str(&body).map_err(|e| WeatherError::Parse(e.to_string()))
}

/// Resolve a location name to (latitude, longitude) via Open-Meteo geocoding.
fn geocode(location: &str, timeout: Duration) -> Result<(f64, f64), WeatherError> {
    let url = format!(
        "https://geocoding-api.open-meteo.com/v1/search?name={}&count=1",
        urlencoding_encode(location)
    );
    let json = http_get_json(&url, timeout)?;
    let first = json
        .get("results")
        .and_then(Value::as_array)
        .and_then(|arr| arr.first())
        .ok_or_else(|| WeatherError::Parse(format!("no geocoding results for {location:?}")))?;
    let lat = first
        .get("latitude")
        .and_then(Value::as_f64)
        .ok_or_else(|| WeatherError::Parse("geocoding result missing latitude".to_string()))?;
    let lon = first
        .get("longitude")
        .and_then(Value::as_f64)
        .ok_or_else(|| WeatherError::Parse("geocoding result missing longitude".to_string()))?;
    Ok((lat, lon))
}

/// Minimal percent-encoding sufficient for a location name query param.
fn urlencoding_encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(b as char)
            }
            b' ' => out.push('+'),
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}

/// Collect one weather report per docs/sdd.md §7.4. Never panics; all
/// failures surface as `WeatherError` for the caller to fold into
/// `weather_error`.
pub fn fetch_weather(cfg: &WeatherConfig, timeout: Duration) -> Result<WeatherReport, WeatherError> {
    let (latitude, longitude, label) = match (cfg.latitude, cfg.longitude) {
        (Some(lat), Some(lon)) => (lat, lon, cfg.location.clone().unwrap_or_default()),
        _ => {
            let location = cfg
                .location
                .as_deref()
                .ok_or_else(|| WeatherError::Config("weather: no location or lat/long set".to_string()))?;
            let (lat, lon) = geocode(location, timeout)?;
            (lat, lon, location.to_string())
        }
    };

    let url = format!(
        "https://api.open-meteo.com/v1/forecast?latitude={latitude}&longitude={longitude}\
         &current=temperature_2m,weather_code\
         &daily=weather_code,temperature_2m_max,temperature_2m_min,precipitation_probability_max\
         &forecast_days={days}&temperature_unit={unit}&timezone=auto",
        days = cfg.forecast_days,
        unit = cfg.temperature_unit,
    );
    let json = http_get_json(&url, timeout)?;
    parse_forecast_json(&json, &label, &cfg.temperature_unit, now_unix())
}

File: app-macos/Sources/WDashboardCore/Weather.swift
// Weather collection per docs/sdd.md §7.4: geocoding + Open-Meteo forecast,
// WMO weather_code -> text mapping, parsed into `WeatherReport` (§5.4).
// Mirrors app-linux/src/weather.rs. JSON parsing and the WMO mapping are
// pure functions, covered by docs/test-vectors/weather-json/ and
// docs/test-vectors/wmo-codes/.

import Foundation

public enum WeatherError: Error, CustomStringConvertible, Sendable {
    case config(String)
    case network(String)
    case parse(String)

    public var description: String {
        switch self {
        case .config(let msg): return "weather config error: \(msg)"
        case .network(let msg): return "weather network error: \(msg)"
        case .parse(let msg): return "weather parse error: \(msg)"
        }
    }
}

/// WMO weather_code -> text mapping, per docs/sdd.md §10.2 (wmo-codes) /
/// docs/test-vectors/wmo-codes/mapping.json. Pure function.
public func wmoDescription(_ code: Int) -> String {
    switch code {
    case 0: return "Clear sky"
    case 1: return "Mainly clear"
    case 2: return "Partly cloudy"
    case 3: return "Overcast"
    case 45: return "Fog"
    case 48: return "Depositing rime fog"
    case 51: return "Light drizzle"
    case 53: return "Moderate drizzle"
    case 55: return "Dense drizzle"
    case 56: return "Light freezing drizzle"
    case 57: return "Dense freezing drizzle"
    case 61: return "Slight rain"
    case 63: return "Moderate rain"
    case 65: return "Heavy rain"
    case 66: return "Light freezing rain"
    case 67: return "Heavy freezing rain"
    case 71: return "Slight snow fall"
    case 73: return "Moderate snow fall"
    case 75: return "Heavy snow fall"
    case 77: return "Snow grains"
    case 80: return "Slight rain showers"
    case 81: return "Moderate rain showers"
    case 82: return "Violent rain showers"
    case 85: return "Slight snow showers"
    case 86: return "Heavy snow showers"
    case 95: return "Thunderstorm"
    case 96: return "Thunderstorm with slight hail"
    case 99: return "Thunderstorm with heavy hail"
    default: return "Unknown"
    }
}

/// WMO weather_code -> SF Symbol name, for use with
/// `.symbolRenderingMode(.multicolor)` (gives each symbol its natural
/// weather coloring — yellow sun, blue rain, etc. — with no extra styling).
/// Pure function.
public func wmoSymbolName(_ code: Int) -> String {
    switch code {
    case 0, 1: return "sun.max.fill"
    case 2: return "cloud.sun.fill"
    case 3: return "cloud.fill"
    case 45, 48: return "cloud.fog.fill"
    case 51, 53, 55: return "cloud.drizzle.fill"
    case 56, 57, 66, 67: return "cloud.sleet.fill"
    case 61, 63, 65: return "cloud.rain.fill"
    case 71, 73, 75, 77: return "cloud.snow.fill"
    case 80, 81, 82: return "cloud.sun.rain.fill"
    case 85, 86: return "cloud.snow.fill"
    case 95: return "cloud.bolt.fill"
    case 96, 99: return "cloud.bolt.rain.fill"
    default: return "cloud.fill"
    }
}

private func asDouble(_ v: Any?) -> Double? {
    (v as? NSNumber)?.doubleValue
}

private func asInt(_ v: Any?) -> Int? {
    (v as? NSNumber)?.intValue
}

/// Parse an Open-Meteo `forecast` JSON response (as decoded by
/// `JSONSerialization`) into a `WeatherReport`. Pure function — covered by
/// docs/test-vectors/weather-json/. `locationLabel`, `temperatureUnit` and
/// `fetchedAt` come from the caller (config/clock), everything else is read
/// from the JSON itself.
public func parseForecastJSON(_ json: Any, locationLabel: String, temperatureUnit: String, fetchedAt: Int64) throws -> WeatherReport {
    guard let root = json as? [String: Any] else {
        throw WeatherError.parse("expected a JSON object at the top level")
    }
    guard let latitude = asDouble(root["latitude"]) else {
        throw WeatherError.parse("missing `latitude`")
    }
    guard let longitude = asDouble(root["longitude"]) else {
        throw WeatherError.parse("missing `longitude`")
    }
    guard let currentObj = root["current"] as? [String: Any] else {
        throw WeatherError.parse("missing `current`")
    }
    guard let temperature = asDouble(currentObj["temperature_2m"]) else {
        throw WeatherError.parse("missing `current.temperature_2m`")
    }
    guard let weatherCode = asInt(currentObj["weather_code"]) else {
        throw WeatherError.parse("missing `current.weather_code`")
    }
    let current = WeatherNow(temperature: temperature, weatherCode: weatherCode, description: wmoDescription(weatherCode))

    var daily: [WeatherDay] = []
    if let dailyObj = root["daily"] as? [String: Any] {
        let dates = (dailyObj["time"] as? [Any]) ?? []
        let codes = dailyObj["weather_code"] as? [Any]
        let tempMaxes = dailyObj["temperature_2m_max"] as? [Any]
        let tempMins = dailyObj["temperature_2m_min"] as? [Any]
        let precip = dailyObj["precipitation_probability_max"] as? [Any]

        for (i, dateVal) in dates.enumerated() {
            guard let date = dateVal as? String else { continue }
            let code = codes.flatMap { arr in arr.indices.contains(i) ? asInt(arr[i]) : nil }
            let tempMax = tempMaxes.flatMap { arr in arr.indices.contains(i) ? asDouble(arr[i]) : nil }
            let tempMin = tempMins.flatMap { arr in arr.indices.contains(i) ? asDouble(arr[i]) : nil }
            // A day missing any required field is dropped rather than
            // failing the whole report, per SDD §7.4 degradation intent.
            guard let code, let tempMax, let tempMin else { continue }
            let precipitationProbabilityMax = precip.flatMap { arr in arr.indices.contains(i) ? asInt(arr[i]) : nil }
            daily.append(WeatherDay(
                date: date,
                tempMax: tempMax,
                tempMin: tempMin,
                weatherCode: code,
                description: wmoDescription(code),
                precipitationProbabilityMax: precipitationProbabilityMax
            ))
        }
    }

    return WeatherReport(
        locationLabel: locationLabel,
        latitude: latitude,
        longitude: longitude,
        fetchedAt: fetchedAt,
        current: current,
        daily: daily,
        temperatureUnit: temperatureUnit
    )
}

private func httpGetJSON(url: URL, timeout: TimeInterval) throws -> Any {
    let config = URLSessionConfiguration.ephemeral
    config.timeoutIntervalForRequest = timeout
    config.timeoutIntervalForResource = timeout
    let session = URLSession(configuration: config)

    let semaphore = DispatchSemaphore(value: 0)
    var resultData: Data?
    var resultError: Error?
    let task = session.dataTask(with: url) { data, _, error in
        resultData = data
        resultError = error
        semaphore.signal()
    }
    task.resume()
    semaphore.wait()

    if let resultError {
        throw WeatherError.network(resultError.localizedDescription)
    }
    guard let data = resultData else {
        throw WeatherError.network("empty response")
    }
    do {
        return try JSONSerialization.jsonObject(with: data)
    } catch {
        throw WeatherError.parse(error.localizedDescription)
    }
}

/// Resolve a location name to (latitude, longitude) via Open-Meteo geocoding.
private func geocode(location: String, timeout: TimeInterval) throws -> (Double, Double) {
    var components = URLComponents(string: "https://geocoding-api.open-meteo.com/v1/search")!
    components.queryItems = [
        URLQueryItem(name: "name", value: location),
        URLQueryItem(name: "count", value: "1"),
    ]
    guard let url = components.url else {
        throw WeatherError.config("invalid location for geocoding: \(location)")
    }
    let json = try httpGetJSON(url: url, timeout: timeout)
    guard let root = json as? [String: Any],
        let results = root["results"] as? [Any],
        let first = results.first as? [String: Any]
    else {
        throw WeatherError.parse("no geocoding results for \"\(location)\"")
    }
    guard let lat = asDouble(first["latitude"]) else {
        throw WeatherError.parse("geocoding result missing latitude")
    }
    guard let lon = asDouble(first["longitude"]) else {
        throw WeatherError.parse("geocoding result missing longitude")
    }
    return (lat, lon)
}

/// Collect one weather report per docs/sdd.md §7.4. Never crashes; all
/// failures surface as `WeatherError` for the caller to fold into
/// `weather_error`.
public func fetchWeather(config: WeatherConfig, timeout: TimeInterval) throws -> WeatherReport {
    let latitude: Double
    let longitude: Double
    let label: String
    if let lat = config.latitude, let lon = config.longitude {
        latitude = lat
        longitude = lon
        label = config.location ?? ""
    } else {
        guard let location = config.location, !location.isEmpty else {
            throw WeatherError.config("weather: no location or lat/long set")
        }
        (latitude, longitude) = try geocode(location: location, timeout: timeout)
        label = location
    }

    var components = URLComponents(string: "https://api.open-meteo.com/v1/forecast")!
    components.queryItems = [
        URLQueryItem(name: "latitude", value: String(latitude)),
        URLQueryItem(name: "longitude", value: String(longitude)),
        URLQueryItem(name: "current", value: "temperature_2m,weather_code"),
        URLQueryItem(name: "daily", value: "weather_code,temperature_2m_max,temperature_2m_min,precipitation_probability_max"),
        URLQueryItem(name: "forecast_days", value: String(config.forecastDays)),
        URLQueryItem(name: "temperature_unit", value: config.temperatureUnit),
        URLQueryItem(name: "timezone", value: "auto"),
    ]
    guard let url = components.url else {
        throw WeatherError.config("failed to build forecast URL")
    }
    let json = try httpGetJSON(url: url, timeout: timeout)
    return try parseForecastJSON(json, locationLabel: label, temperatureUnit: config.temperatureUnit, fetchedAt: Int64(Date().timeIntervalSince1970))
}
